tools/setup_services.py

```diff
@@ -1,15 +1,396 @@
+        return tomli.load(f)
+
+
+def get_plugin_compatible_version(plugin_config: dict):
+    # if the plugin specifies a "tool.<name>" section with
+    # a version requirement, retrive it
+    if "tool" not in plugin_config:
+        log("Plugin config does not contain 'tool' section")
+        return None, None
+    # there should only be one key in tool, which is the name of the package
+    # for which we are installing the plugin
+    if len(plugin_config["tool"]) != 1:
+        log("Plugin config must have exactly one 'tool' section")
+        return None, None
+
+    tool_name = next(iter(plugin_config["tool"]))
+    tool_version_requirement = plugin_config["tool"][tool_name].get("version", None)
+
+    if not tool_version_requirement:
+        log(f"Plugin config for '{tool_name}' does not specify a version")
+        return None, None
+
+    return tool_name, tool_version_requirement
+
+
+def validate_version(version_string, requirement_string):
+    """
+    Validate a version against a version requirement.
+
+    Args:
+        version_string (str): The version to check (e.g., "1.2.3")
+        requirement_string (str): The requirement specification (e.g., ">=1.0.0", "==1.2.3", ">=1.0,<2.0")
+
+    Returns:
+        bool: True if version satisfies the requirement, False otherwise
+    """
+    try:
+        v = version.parse(version_string)
+        spec = SpecifierSet(requirement_string)
+        return v in spec
+    except Exception as e:
+        log(f"Error validating version: {e}")
+        return False
+
+
+def validate_plugin_compatibility(plugin_name: str, plugin_path: str):
+    plugin_config = read_plugin_config(plugin_path)
+    name, version_requirement = get_plugin_compatible_version(plugin_config)
+    if not (name and version_requirement):
+        return True
+    try:
+        mod = __import__(name)
+        installed_version = mod.__version__
+
+        if not validate_version(installed_version, version_requirement):
+            log(
+                f"Plugin {plugin_name} is incompatible: required {version_requirement}, found {installed_version}. Skipping."
+            )
+            return False
+    except ImportError:
+        log(
+            f"Could not find package {name} which plugin {plugin_name} depends on. Skipping."
+        )
+        return False
+
+    if plugin_config["project"]["name"] != plugin_name:
+        log(
+            f"Plugin {plugin_name} has a different name in its config ({plugin_config['project']['name']}). Skipping."
+        )
+        return False
+
+    return True
+
+
+def run_git_command(args, plugin_path, plugin_name):
+    """
+    Run a git command in the specified plugin path.
+
+    Args:
+        args (list): Git command, e.g. ['checkout', 'main']
+        plugin_path (str): Directory where the command runs
+        plugin_name (str): Used in error logs
+
+    Returns:
+        stdout_lines (list of str): Output lines from stdout
```

Comment:
Why does stdout get split into lines, but stderr is provided as a str?

Reply:
I think @thomasculino needed to parse multiple lines for some commands, but we should standardize that. We'll talk about it